### ISSUE TITLE
Validator option "--limit-ledger-size" does not work as intended

### ISSUE THREAD

**Issue by sagar-solana** [solana-labs/solana#6521]
#### Problem

`--limit-ledger-size` doesn't seem to be working as strictly as we hoped. There appears to be some ballooning of disk usage over time. 

#### Proposed Solution

Triage and fix.
This is likely happening because of the size of some slowly updating columns in Rocks. We should be able to force flush to disk every so often so that the delete commands get committed instead of just hanging around until the slow columns cause a flush.

**Comment by carllin** [comment]
It might be because we are hanging onto WAL due to smaller CF's not being flushed. We should see if it works with WAL turned off. We can also try setting the `max_total_wal_size` here: https://github.com/facebook/rocksdb/wiki/Write-Ahead-Log

**Comment by mvines** [comment]
I'm seeing the steady state ledger size with `--limit-ledger-size` to be ~250GB on the TdS cluster.  This is not super but also not terrible.  So while we should resolve this issue still, I think we don't need it in 0.21

**Comment by mvines** [comment]
@sagar-solana - do you want this one, or hand off to @sunnygleason?

**Comment by sagar-solana** [comment]
I think still depends directly on Sunny's compaction work. Compacting the ledger makes limit ledger size(all the delete calls) work.

**Comment by sagar-solana** [comment]
@mvines soooo.. I just found a profoundly dumb bug in the purge_slots logic. Which probably explains why this flag was not doing anything. 

So I have a feeling https://github.com/solana-labs/solana/pull/7515 will indeed fix the flag.
I'll give it a shot tomorrow and check to make sure. 
Then compaction should be the final push. 

TL;DR: I was deleting the columns with a bunch of logical `&&`s. I forgot that rust will short circuit the operation if any part returns false. Switched to `&` now so that it doesn't lazily eval the expression. :man_facepalming:
